### DISCUSSION TITLE
CODEOWNERS works with users but not teams

### DISCUSSION THREAD

**Discussion by notmessenger** [community/community#22751] [New to GitHub]
<p>In <a href="https://github.com/ciena-blueplanet/ciena-devops-testbed" rel="nofollow noopener">https://github.com/ciena-blueplanet/ciena-devops-testbed</a> when using the CODEOWNERS entries of</p>
<pre><code>* @notmessenger @juwara0 @job13er

README.md @notmessenger @juwara0
</code></pre>
<p>and making a PR changing just the <em>.THROWAWAY</em> file all three users (<a class="mention" href="/notmessenger">@notmessenger</a> <span class="mention">@juwara0</span> <a class="mention" href="/job13er">@job13er</a>) are added as reviewers and it only takes one approver for the PR to be mergeable (<a href="https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/45" rel="nofollow noopener">https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/45</a>)</p>
<p>When the <em>.THROWAWAY</em> file and the <em>README.md</em> file are changed it requires one of the three users (<a class="mention" href="/notmessenger">@notmessenger</a> <span class="mention">@juwara0</span> <a class="mention" href="/job13er">@job13er</a>) as well as one of the two (<a class="mention" href="/notmessenger">@notmessenger</a> <span class="mention">@juwara0</span>) to be mergeable (<a href="https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/46" rel="nofollow noopener">https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/46</a>)</p>
<p>This is exactly what I expected to happen per <a href="https://help.github.com/articles/about-codeowners/" rel="nofollow noopener">https://help.github.com/articles/about-codeowners/</a></p>
<p>However when I attempt to do similar via using teams things get all messed up and I’m not sure why.</p>
<p>I created the <a href="https://github.com/orgs/ciena-blueplanet/teams/reviewers-tooling/members" rel="nofollow noopener">https://github.com/orgs/ciena-blueplanet/teams/reviewers-tooling/members</a> team and added <a class="mention" href="/notmessenger">@notmessenger</a>, <span class="mention">@juwara0</span>, and <a class="mention" href="/job13er">@job13er</a> to it.</p>
<p>I created the <a href="https://github.com/orgs/ciena-blueplanet/teams/owners-architecture" rel="nofollow noopener">https://github.com/orgs/ciena-blueplanet/teams/owners-architecture</a> team and added <a class="mention" href="/notmessenger">@notmessenger</a> and <span class="mention">@juwara0</span> to it.</p>
<p>I then used these CODEOWNERS entries:</p>
<pre><code>* @ciena-blueplanet/reviewers-tooling

README.md @ciena-blueplanet/owners-architecture
</code></pre>
<p>and got this result for only the <em>.THROWAWAY</em> file changed (<a href="https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/45" rel="nofollow noopener">https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/45</a>):</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/2ef38872494f6535b69bb1d1d957a1339329cbdf.png" data-download-href="/uploads/short-url/6HlJiHZxKaPioJ80oukLJIrlJEP.png?dl=1" title="37438102-92eabe4e-27be-11e8-97c7-2f1f70b0223f" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/435544/181073004-620be680-76c4-4a5d-a0d8-449b6fe94166.png" alt="37438102-92eabe4e-27be-11e8-97c7-2f1f70b0223f" data-base62-sha1="6HlJiHZxKaPioJ80oukLJIrlJEP" width="525" height="499" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/2ef38872494f6535b69bb1d1d957a1339329cbdf_2_525x499.png, https://aws1.discourse-cdn.com/github/original/1X/2ef38872494f6535b69bb1d1d957a1339329cbdf.png 1.5x, https://aws1.discourse-cdn.com/github/original/1X/2ef38872494f6535b69bb1d1d957a1339329cbdf.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/2ef38872494f6535b69bb1d1d957a1339329cbdf_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">37438102-92eabe4e-27be-11e8-97c7-2f1f70b0223f</span><span class="informations">781×743 73.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>and this result for the <em>.THROWAWAY</em> file and <em>README.md</em> file changed (<a href="https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/46" rel="nofollow noopener">https://github.com/ciena-blueplanet/ciena-devops-testbed/pull/46</a>):</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/99de3e2ae04698407c27a4f2b869f4dc83924a06.png" data-download-href="/uploads/short-url/lXbcM8whTkpiurSwPDwcdFA9e2W.png?dl=1" title="37438108-9d8c9c28-27be-11e8-8a6d-d5c1d16a534d" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/435544/181073005-3a2957b8-34db-4b40-9460-2326acba8a49.png" alt="37438108-9d8c9c28-27be-11e8-8a6d-d5c1d16a534d" data-base62-sha1="lXbcM8whTkpiurSwPDwcdFA9e2W" width="528" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/99de3e2ae04698407c27a4f2b869f4dc83924a06_2_528x500.png, https://aws1.discourse-cdn.com/github/original/1X/99de3e2ae04698407c27a4f2b869f4dc83924a06.png 1.5x, https://aws1.discourse-cdn.com/github/original/1X/99de3e2ae04698407c27a4f2b869f4dc83924a06.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/99de3e2ae04698407c27a4f2b869f4dc83924a06_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">37438108-9d8c9c28-27be-11e8-8a6d-d5c1d16a534d</span><span class="informations">778×736 75.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Have I set the teams and/or CODEOWNERS up correctly?</p>
<p>Sorry that I only have screenshots for the failed state - I am using the same repo to work through these issues so I made changes to the files for different scenarios I’m testing. It may even be likely that the current content of the CODEOWNERS file are different than what I’ve pasted here as I’m still making changes but the content pasted here is accurate and I will not be making any changes to the teams I have setup so they can be looked at.</p>
<p>You can ignore the double travis-ci required status checks - I know why there are two and it is unrelated to this issue.</p>
<p>I also now realize that my screen captures did not capture the reviewers that were assigned.  For PR 45 it did not assign everyone in the <a href="https://github.com/orgs/ciena-blueplanet/teams/reviewers-tooling/members" rel="nofollow noopener">https://github.com/orgs/ciena-blueplanet/teams/reviewers-tooling/members</a> team as I expected, and like it does when individual users are listed.  For PR 46 it assigned no one at all.</p>

**Comment by notmessenger** [comment, marked as answer]
<p>I have received this reply from the GitHub Support Staff which I will give a try and report back on the status of:</p>
<pre><code>It looks like you've set up the CODEOWNERS file correctly, but I believe the issue you're having is down to team permissions.

Teams are only requested for a review on a pull request when they have write access to the repository on the team itself. It's not enough for all the members to have write access, you need to grant write access to the actual team.

https://help.github.com/articles/managing-team-access-to-an-organization-repository/

Once you do this all future pull requests that involve files owned by that team should be automatically requested for a review.
</code></pre>

**Comment by notmessenger** [comment]
<p>This is exactly what the problem was.</p>

**Comment by VinceSJ** [comment]
<p>I’m having a similar issue, but I’m pretty sure mine is not being caused by team permissions. Here’s my CODEOWNERS file (with org and team name anonymized):</p>
<pre><code># This file is used to automatically request reviews on PRs.

# These owners will be the default owners for everything in
# the repo, unless a match on a later line takes precedence.
# "@MyOrganization/MyTeam" below requests review from all members of MyTeam.
* @MyOrganization/MyTeam

# It's also possible to add notifications based on language
# or location in repo, etc.
# More info at docs: https://help.github.com/articles/about-codeowners/
</code></pre>
<p>However, when I make a PR in a repo with that CODEOWNERS file, no members of the team are notified for review. After seeing <a class="mention" href="/notmessenger">@notmessenger</a>’s message from GitHub Support saying that the team needs write permissions for the repo, I verified the team had write permission for the repo, but subsequent PRs still do not kick off a review request by team members.</p>
<p>Am I missing something else? I can replace the “<span class="mention">@MyOrganization</span>/MyTeam” part with a list of all the team users and reviews are automatically requested, but I want to be able to do this on a team basis, not on a user-by-user basis.</p>
<p>The only other two confounding factors I can think of are these:</p>
<ol>
<li>The team name’s capitalization is kinda funky. Clicking through the org’s team list, I can see the team capitalized in some places (MYTEAM), lowercaps in others (myteam). Asking our admin to hit the GitHub API, I found that the team’s “name” field is <code>MYTEAM</code>, whereas its “slug” field is <code>myteam</code>. But at this point I’ve tried using both capitalization styles in my CODEOWNERS file and neither has worked, so that shouldn’t be it.</li>
<li>In the org, the team is marked as a “Secret” team. (Mousing over the “Secret” box, a tooltip appears and says “Only visible to its members.”) If it’s necessary, I can talk to admin and see if we can remove the team’s “Secret” designation, but I’d rather avoid that.</li>
</ol>

**Comment by VinceSJ** [comment]
<p>Following up – I did speak to GitHub support, and the issue was the “Secret” status for the team. CODEOWNER notifications to members of a “Secret” team (at least as of late October) just aren’t supported.</p>
<p>Luckily my team was able to switch over to “Visible” (the opposite of Secret status), at which point the CODEOWNERS notifications worked exactly as expected. If you need to keep the Secret status though, I think there’s currently no way to get the auto notifications as well.</p>

**Reply by evencheng-rl** [reply]
Yes! I have the same set up. Once I move the team to be `Visible`, the auto-assign is happening.

Also see this description next to the `Visible` setting
```
A visible team can be seen and [@mentioned](https://github.blog/2012-05-09-introducing-team-mentions/) by every member of this organization.
```

**Comment by jfairley** [comment]
<p>I’m having this problem also.</p>
<ul>
<li>My team is “visible”.</li>
<li>My team has “Write” access to the repository.</li>
</ul>
<p>I’m not sure what else to try.</p>
<ul>
<li>My organization is private. Could that be a problem?</li>
<li>My organization name is not all lowercase. Could case sensitivity be a problem?
<ul>
<li>A note here, I’ve tried both “AdCellerant” and “adcellerant” variants with no luck. <em>If</em> this is the problem, it likely resides within <a href="http://github.com" rel="noopener nofollow ugc">github.com</a>.</li>
</ul>
</li>
</ul>
<p>I have this issue both in my <code>CODEOWNERS</code> file and when trying to use <code>hub</code>.</p>
<p>BTW, the team I’m trying to hook up is <code>@AdCellerant/app</code>.</p>
<h3>CODEOWNERS</h3>
<p>I know this worked in the past. It stopped working the day I added <code>app/renovate-approve</code>. Since then, it doesn’t, even when I remove <code>app/renovate-approve</code>. I’m stumped.</p>
<h3>Hub</h3>
<p>I know this part is unrelated but I’m hoping it’s helpful…</p>
<p>I’m trying to assign team reviewers via <code>repo-sync/pull-request@v2</code> which uses <code>hub</code> under the hood.</p>
<p>This is the API call that fails. (Hub executes the POST.)</p>
<pre><code class="lang-auto">&gt; POST api.github.com/repos/AdCellerant/ui-marketing-backend/pulls/461/requested_reviewers
&gt; Authorization: token [REDACTED]
&gt; Accept: application/vnd.github.v3+json;charset=utf-8
{"reviewers":[],"team_reviewers":["app"]}
&lt; HTTP 422
{"message":"Validation Failed","errors":["Could not resolve to a node with the global id of 'MDQ6VGVhbTMyNjk0NDI='."],"documentation_url":"docs.github.com/rest/reference/pulls#request-reviewers-for-a-pull-request"}
Error requesting reviewer: Unprocessable Entity (HTTP 422)
Could not resolve to a node with the global id of 'MDQ6VGVhbTMyNjk0NDI='.
</code></pre>

**Reply by MPV** [reply]
See this bug:
- `Actions and Apps Can't Request Review From A Team` \
   https://github.com/orgs/community/discussions/66049

**Comment by gardnerjr** [comment]
<p>Ours just started having this issue as well.  We switched from having a lot of members to having teams with members (at at least the top level team i’m on is definitely visible), and now our PRs don’t have CODEOWNERS properly assigned either.</p>

**Comment by amdove** [comment]
Was there ever a fix to this? Seems that granting the team write access wasn't enough for some to fix this behavior.

**Comment by oerp-odoo** [comment]
Having same problem. Team has write access, is also `visible` not `secret`, but still no one is assigned from a team.

**Comment by chrisinmtown** [comment]
The upside of a team in codeowners is, no maintenance in the repo when the team members change. And email notifications to request reviews on pull requests go out as before. The downside appears to be that the Github search for pull requests page does not evaluate the clause `involves:@me` to true for PRs where only the *team* is named, and the members of the team have not yet commented or posted a review. This seems like a relatively easy thing to fix. How to request that?

**Comment by scott-brenstuhl-infinitusai** [comment]
If nothing else is working, double check where your file is. I accidently had mine in `.github/workflows/CODEOWNERS`  instead of `.github/CODEOWNERS`. Github still validates that it is a valid CODEOWNER file which... fair, but I think did lead me to assume it was also in the correct place lol.

**Comment by ay-azara** [comment]
Just to aggregate some of the requirements for newcomers
* Team must be set to "visible" to be `@mentioned` for reviews
* Team must have write access to the repository
* Team name must be in exact case as CODEOWNERS file is case-sensitive
* Team name must include the organization ala `@octo-org/codeowners-team`
* CODEOWNERS is not additive like you'd imagine, it's a switch that breaks at the first match starting from the bottom.
  In other words, if there is a match somewhere below your Team it will not assign your team
